File: textarc/src/main/java/com/a_lab/textarc/TextArc.java
package com.a_lab.textarc;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.Nullable;
import androidx.core.content.res.ResourcesCompat;

public class TextArc extends View {

    private int radius = 320;
    //private int startAngle = Integer.MIN_VALUE;
    //private int sweepAngle = Integer.MIN_VALUE;
    private int centerAngle =-90;
    private float textSize = getResources().getDisplayMetrics().density * 16;
    private String text = "";

    private int textColor = Color.WHITE;
    private Typeface fontFamily;

    private Paint paintText = new Paint(Paint.ANTI_ALIAS_FLAG);
    private float offset = 0;

    public TextArc(Context context) {
        super(context);
    }

    public TextArc(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        TypedArray typedArray = getContext().obtainStyledAttributes(attrs, R.styleable.TextArc);

        radius = typedArray.getDimensionPixelSize(R.styleable.TextArc_radius, radius);
        //startAngle = typedArray.getInteger(R.styleable.TextArc_start_angle, startAngle);
        //sweepAngle = typedArray.getInteger(R.styleable.TextArc_sweep_angle, sweepAngle);
        centerAngle = typedArray.getInteger(R.styleable.TextArc_center_angle, centerAngle);
        textSize = typedArray.getDimensionPixelSize(R.styleable.TextArc_text_size, (int) textSize);

        String text = typedArray.getString(R.styleable.TextArc_text);
        if (text != null)
            this.text = text;

        int colorRes = typedArray.getColor(R.styleable.TextArc_text_color, Color.WHITE);
        if (colorRes != -1)
            textColor = colorRes;

        int fontRes = typedArray.getResourceId(R.styleable.TextArc_font_family, -1);
        if (fontRes != -1)
            fontFamily = ResourcesCompat.getFont(context, fontRes);

        typedArray.recycle();
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();

        offset = textSize * 0.75f;

        ViewGroup.LayoutParams lp = getLayoutParams();
        lp.width = radius > 0 ? (int) (radius * 2 + offset * 2) : 0;
        lp.height = radius > 0 ? (int) (radius * 2 + offset * 2) : 0;
        requestLayout();

        //Text color
        paintText.setColor(textColor);
        paintText.setTypeface(fontFamily);
        paintText.setTextSize(textSize);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {

    }

    @Override
    protected void onDraw(Canvas canvas) {
        float textWidth = paintText.measureText(text);
        int circumference = (int) (2 * Math.PI * radius);   //Длина окружности

        int textAngle = (int) (textWidth * 360 / circumference);   //Угол занимаемый текстом
        int startAngle = centerAngle - (textAngle / 2);

        RectF oval = new RectF(offset, offset,radius * 2 + offset,radius * 2 + offset);
        Path pathArc = new Path();
        pathArc.addArc(oval, startAngle, 350);

        canvas.drawTextOnPath(text, pathArc, 0, 0, paintText);
    }


    //Устанавливаем текст
    public void setText(String text) {
        this.text = text;
        invalidate();
    }

    public void setRadius(int radius) {
        this.radius = radius;
        invalidate();
    }

    //Устанавливаем текст
    public void setCenterAngle(int centerAngle) {
        this.centerAngle = centerAngle;
        invalidate();
    }

    //Устанавливаем цвет текста
    public void setTextColor(int textColor) {
        this.textColor = textColor;
        invalidate();
    }

    //Устанавливаем цвет текста
    public void setTextSize(int textSize) {
        this.textSize = textSize;
        invalidate();
    }

    //Устанавливаем цвет текста
    public void setFontFamily(Typeface fontFamily) {
        this.fontFamily = fontFamily;
        invalidate();
    }

}

File: textarc/src/main/java/com/a_lab/textarc/TextArc.kt
package com.a_lab.textarc

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import androidx.core.content.res.ResourcesCompat

/**
 * Text arc.
 */
class TextArc @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(
    context,
    attrs,
    defStyleAttr
) {
    private val rect = RectF()
    private val path = Path()
    private var radius = 320
    //private int startAngle = Integer.MIN_VALUE;
    //private int sweepAngle = Integer.MIN_VALUE;
    private var centerAngle = -90
    private var textSize: Int = (resources.displayMetrics.density * 16).toInt()
    private var text = ""

    private var textColor = Color.WHITE
    private var fontFamily: Typeface? = null

    private val paintText = Paint(Paint.ANTI_ALIAS_FLAG)
    private var offset = 0

    init {
        val typedArray = context.obtainStyledAttributes(attrs, R.styleable.TextArc)

        radius = typedArray.getDimensionPixelSize(R.styleable.TextArc_radius, radius)

        //startAngle = typedArray.getInteger(R.styleable.TextArc_start_angle, startAngle);
        //sweepAngle = typedArray.getInteger(R.styleable.TextArc_sweep_angle, sweepAngle);
        centerAngle = typedArray.getInteger(R.styleable.TextArc_center_angle, centerAngle)
        textSize = typedArray.getDimensionPixelSize(R.styleable.TextArc_text_size, textSize)

        text = typedArray.getString(R.styleable.TextArc_text) ?: ""
        textColor = typedArray.getColor(R.styleable.TextArc_text_color, Color.WHITE)

        fontFamily = typedArray.getResourceId(R.styleable.TextArc_font_family, -1).let {
            if (it == -1) {
                null
            } else {
                ResourcesCompat.getFont(context, it)
            }
        }

        typedArray.recycle();
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()

    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        offset = (textSize * 0.75f).toInt()

        val width = if (radius > 0) radius * 2 + offset * 2 else 0
        val height = if (radius > 0) radius * 2 + offset * 2 else 0

        setMeasuredDimension(width, height)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        paintText.color = textColor
        paintText.typeface = fontFamily
        paintText.textSize = textSize.toFloat()

        super.onLayout(changed, left, top, right, bottom)
    }

    override fun onDraw(canvas: Canvas) {
        val textWidth = paintText.measureText(text)
        val circumference = (2 * Math.PI * radius).toFloat() //Circumference length
        val textAngle = textWidth * 360 / circumference //Угол занимаемый текстом
        val startAngle = centerAngle - textAngle / 2
        val oval = rect.apply {
            left = offset.toFloat()
            top = offset.toFloat()
            right = (radius * 2 + offset).toFloat()
            bottom = (radius * 2 + offset).toFloat()
        }
        path.reset()
        path.addArc(oval, startAngle, 350f)
        canvas.drawTextOnPath(text, path, 0f, 0f, paintText)
    }


    /**
     * Sets text.
     * @param text Text.
     */
    fun setText(text: String?) {
        this.text = text ?: ""
        invalidate()
    }

    /**
     * Sets radius.
     * @param radius Radius.
     */
    fun setRadius(radius: Int) {
        this.radius = radius
        invalidate()
    }

    /**
     * Sets center angle.
     * @param centerAngle Center angle.
     */
    fun setCenterAngle(centerAngle: Int) {
        this.centerAngle = centerAngle
        invalidate()
    }

    /**
     * Sets text color.
     * @param textColor Text color.
     */
    fun setTextColor(textColor: Int) {
        this.textColor = textColor
        invalidate()
    }

    /**
     * Sets text size.
     * @param textSize Text size.
     */
    fun setTextSize(textSize: Int) {
        this.textSize = textSize.toFloat().toInt()
        invalidate()
    }

    /**
     * Sets font family.
     * @param fontFamily Font family.
     */
    fun setFontFamily(fontFamily: Typeface?) {
        this.fontFamily = fontFamily
        invalidate()
    }

}